File: scraper/scraper/normalize.py
"""
normalize_lead and content_hash — TRD §3.1

CRITICAL: This implementation must produce bit-for-bit identical output to
any Go re-computation in the gateway. The algorithm is:
  1. Lowercase entire string
  2. Remove HTML tags
  3. Remove non-alphanumeric / non-space characters  ([^a-z0-9 ])
  4. Collapse all whitespace to single spaces and strip()
  5. Take the first 300 characters

content_hash = SHA-256(canonical_url + normalize_lead(text))
"""

import hashlib
import re
from html.parser import HTMLParser


class _HTMLStripper(HTMLParser):
    """Minimal HTML stripper that collects text nodes only."""

    def __init__(self) -> None:
        super().__init__()
        self._parts: list[str] = []

    def handle_data(self, data: str) -> None:
        self._parts.append(data)

    def result(self) -> str:
        return "".join(self._parts)


def _strip_html(text: str) -> str:
    s = _HTMLStripper()
    s.feed(text)
    return s.result()


def normalize_lead(text: str) -> str:
    """
    Canonical normalization per TRD §3.1.
    Returns the first 300 chars of the normalized string.
    """
    text = text.lower()
    text = _strip_html(text)
    text = re.sub(r"[^a-z0-9\s]", "", text)  # keep lowercase alphanum + any whitespace
    text = re.sub(r"\s+", " ", text)           # collapse all whitespace (tabs, newlines) to one space
    text = text.strip()
    return text[:300]


def content_hash(canonical_url: str, lead_text: str) -> str:
    """
    SHA-256(canonical_url + normalize_lead(lead_text)) as a lowercase hex digest.
    The canonical_url is concatenated raw (no separator) before hashing.
    """
    payload = canonical_url + normalize_lead(lead_text)
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()

File: gateway/internal/feed/normalize.go
package feed

import (
	"crypto/sha256"
	"fmt"
	"regexp"
	"strings"

	"golang.org/x/net/html"
)

// normalizeRegexp removes any character that is not a-z, 0-9, or whitespace.
// Must stay in sync with Python: re.sub(r"[^a-z0-9\s]", "", text)
var normalizeRegexp = regexp.MustCompile(`[^a-z0-9\s]`)

// whitespaceRegexp collapses all runs of whitespace (space, tab, newline) to a single space.
// Must stay in sync with Python: re.sub(r"\s+", " ", text)
var whitespaceRegexp = regexp.MustCompile(`\s+`)

// NormalizeLead implements the canonical normalization algorithm from TRD §3.1.
// Output must be bit-for-bit identical to the Python scraper's normalize_lead().
//
// Steps:
//  1. Lowercase
//  2. Strip HTML tags (extract text nodes only)
//  3. Remove non-alphanumeric / non-whitespace characters
//  4. Collapse all whitespace to single spaces and strip
//  5. Take first 300 characters (by rune, not byte)
func NormalizeLead(text string) string {
	text = strings.ToLower(text)
	text = stripHTML(text)
	text = normalizeRegexp.ReplaceAllString(text, "")
	text = whitespaceRegexp.ReplaceAllString(text, " ")
	text = strings.TrimSpace(text)

	runes := []rune(text)
	if len(runes) > 300 {
		runes = runes[:300]
	}
	return string(runes)
}

// ContentHash computes SHA-256(canonical_url + NormalizeLead(leadText)) per TRD §3.1.
func ContentHash(canonicalURL, leadText string) string {
	payload := canonicalURL + NormalizeLead(leadText)
	sum := sha256.Sum256([]byte(payload))
	return fmt.Sprintf("%x", sum)
}

// stripHTML extracts text nodes from an HTML fragment, discarding all tags.
func stripHTML(s string) string {
	doc, err := html.Parse(strings.NewReader(s))
	if err != nil {
		// Fallback: return as-is; downstream normalization will handle it.
		return s
	}
	var b strings.Builder
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if n.Type == html.TextNode {
			b.WriteString(n.Data)
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(doc)
	return b.String()
}
